
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>coverage &#8212; Counter RNAseq Window 1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="heatmap" href="heatmap.html" />
    <link rel="prev" title="wig" href="wig.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="heatmap.html" title="heatmap"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="wig.html" title="wig"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="coverage">
<span id="id1"></span><h1>coverage<a class="headerlink" href="#coverage" title="Permalink to this headline">¶</a></h1>
<p><em>coverage</em> module contain several functions which allow to get the coverages from data input, a
<a class="reference internal" href="wig.html#craw.wig.Genome" title="craw.wig.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">craw.wig.Genome</span></code></a> object or a <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.AlignmentFile</span></code>.</p>
<p>There is 2 kind of functions:</p>
<ul class="simple">
<li><p>the functions to get coverage from input.</p></li>
<li><p>the functions to process the coverage.</p></li>
</ul>
<div class="section" id="functions-to-get-coverage">
<h2>Functions to get coverage<a class="headerlink" href="#functions-to-get-coverage" title="Permalink to this headline">¶</a></h2>
<p>These low level functions are not aimed to be called directly.
They are called inside function which process the coverages.</p>
<div class="section" id="get-raw-bam-coverage">
<h3>get_raw_bam_coverage<a class="headerlink" href="#get-raw-bam-coverage" title="Permalink to this headline">¶</a></h3>
<p>Get coverage from <cite>pysam</cite> for reference (<em>chromosome</em>) for an interval of positions, a quality on both strand.
and convert the coverage return by <cite>pysam</cite>. A score on each position for each base (ACGT)) in a global coverage for this
position.</p>
<p>This function is called for each entry of the annotation file.</p>
</div>
<div class="section" id="get-raw-wig-coverage">
<h3>get_raw_wig_coverage<a class="headerlink" href="#get-raw-wig-coverage" title="Permalink to this headline">¶</a></h3>
<p>Get coverage from <a class="reference internal" href="wig.html#craw.wig.Genome" title="craw.wig.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">craw.wig.Genome</span></code></a> instance for reference (<em>chromosome</em>) for an interval of positions, on both strand.
The quality parameter is here just to have the same signature as get_bam_coverage but will be ignores .</p>
<p>This function is called for each entry of the annotation file.</p>
</div>
<div class="section" id="get-raw-coverage-function">
<h3>get_raw_coverage_function<a class="headerlink" href="#get-raw-coverage-function" title="Permalink to this headline">¶</a></h3>
<p>Allow to choose the right <em>get_raw_(*)_coverage</em> in function of the data input type
(<a class="reference internal" href="wig.html#craw.wig.Genome" title="craw.wig.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">craw.wig.Genome</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.AlignmentFile</span></code>)</p>
<span class="target" id="module-craw.coverage"></span><dl class="function">
<dt id="craw.coverage.get_raw_wig_coverage">
<code class="descclassname">craw.coverage.</code><code class="descname">get_raw_wig_coverage</code><span class="sig-paren">(</span><em>genome</em>, <em>annot_entry</em>, <em>start</em>, <em>stop</em>, <em>qual_thr=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/coverage.html#get_raw_wig_coverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.coverage.get_raw_wig_coverage" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome</strong> (<a class="reference internal" href="wig.html#craw.wig.Genome" title="craw.wig.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">craw.wig.Genome</span></code></a> object) – The genome which store all coverages.</p></li>
<li><p><strong>annot_entry</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">annotation.Entry</span></code> object) – an entry of the annotation file</p></li>
<li><p><strong>start</strong> (<em>int</em>) – The position to start to compute the coverage(coordinates are 0-based, start position is included).</p></li>
<li><p><strong>stop</strong> (<em>int</em>) – The position to stop to compute the coverage (coordinates are 0-based, stop position is excluded).</p></li>
<li><p><strong>qual_thr</strong> (<em>None</em>) – this parameter is not used, It’s here to have the same api as get_bam_coverage.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the coverage (all bases)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple of 2 list containing int or float</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="craw.coverage.get_raw_bam_coverage">
<code class="descclassname">craw.coverage.</code><code class="descname">get_raw_bam_coverage</code><span class="sig-paren">(</span><em>sam_file</em>, <em>annot_entry</em>, <em>start</em>, <em>stop</em>, <em>qual_thr=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/coverage.html#get_raw_bam_coverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.coverage.get_raw_bam_coverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the coverage for a region position by position on each strand</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sam_file</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.AlignmentFile</span></code> object.) – the samfile openend with pysam</p></li>
<li><p><strong>annot_entry</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">annotation.Entry</span></code> object) – an entry of the annotation file</p></li>
<li><p><strong>start</strong> (<em>positive int</em>) – The position to start to compute the coverage(coordinates are 0-based, start position is included).</p></li>
<li><p><strong>stop</strong> (<em>positive int</em>) – The position to start to compute the coverage (coordinates are 0-based, stop position is excluded).</p></li>
<li><p><strong>qual_thr</strong> (<em>int</em>) – The quality threshold</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the coverage (all bases)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple of 2 list containing int</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="craw.coverage.get_raw_coverage_function">
<code class="descclassname">craw.coverage.</code><code class="descname">get_raw_coverage_function</code><span class="sig-paren">(</span><em>input</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/coverage.html#get_raw_coverage_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.coverage.get_raw_coverage_function" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>input</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">wig.Genome</span></code> or <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.calignmentfile.AlignmentFile</span></code> object) – the input either a samfile (see pysam library) or a genome build from a wig file (see wig module)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>get_wig_coverage or get_bam_coverage according the type of input</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>function</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>RuntimeError</strong> – when input is not instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.calignmentfile.AlignmentFile</span></code> or <code class="xref py py-class docutils literal notranslate"><span class="pre">wig.Genome</span></code></p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="functions-to-process-coverages">
<h2>Functions to process coverages<a class="headerlink" href="#functions-to-process-coverages" title="Permalink to this headline">¶</a></h2>
<p>These functions guess the right <em>get_raw_(*)_coverage</em> in function of the data input and pass
it to a post processing function.</p>
<p>all functions returned have the same API</p>
<p>3 parameters as input</p>
<ul class="simple">
<li><p><strong>annot_entry</strong>: an entry of the annotation file.</p></li>
<li><p><strong>start</strong>: The position to start to compute the coverage(coordinates are 0-based, start position is included).</p></li>
<li><p><strong>stop</strong>: The position to stop to compute the coverage (coordinates are 0-based, stop position is excluded).</p></li>
</ul>
<p>and</p>
<ul class="simple">
<li><p><strong>return</strong>: a tuple of two list or tuple containing in this order the coverages on the forward strand then the
coverages on the reverse strand.</p></li>
</ul>
<p>These architecture allow to combine easily the different get_raw_coverage function with the different post-processing.
For instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="n">get_coverage</span> <span class="o">=</span> <span class="n">get_padded_coverage</span><span class="p">(</span><span class="n">bam</span><span class="p">,</span> <span class="n">max_left</span><span class="p">,</span> <span class="n">max_right</span><span class="p">,</span> <span class="n">qual_thr</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
<span class="n">forward</span><span class="p">,</span> <span class="n">reverse</span> <span class="o">=</span> <span class="n">get_coverage</span><span class="p">(</span><span class="n">annot_entry</span><span class="p">,</span> <span class="mi">10</span> <span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
<p>or</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wig</span> <span class="o">=</span> <span class="n">wig_parser</span><span class="o">.</span><span class="n">parse</span><span class="p">()</span>
<span class="n">get_coverage</span> <span class="o">=</span> <span class="n">get_resized_coverage</span><span class="p">(</span><span class="n">wig</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
<span class="n">forward</span><span class="p">,</span> <span class="n">reverse</span> <span class="o">=</span> <span class="n">get_coverage</span><span class="p">(</span><span class="n">annot_entry</span><span class="p">,</span> <span class="mi">10</span> <span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
<span class="target" id="module-craw.coverage"></span><dl class="function">
<dt id="craw.coverage.padded_coverage_maker">
<code class="descclassname">craw.coverage.</code><code class="descname">padded_coverage_maker</code><span class="sig-paren">(</span><em>input_data</em>, <em>max_left</em>, <em>max_right</em>, <em>qual_thr=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/coverage.html#padded_coverage_maker"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.coverage.padded_coverage_maker" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_data</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">wig.Genome</span></code> or <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.AlignmentFile</span></code> object) – the input either a samfile (see pysam library) or a genome build from a wig file (see wig module)</p></li>
<li><p><strong>max_left</strong> (<em>int</em>) – The highest number of base before the reference position to take in account.</p></li>
<li><p><strong>max_right</strong> (<em>int</em>) – The highest number of base after  the reference position to take in account.</p></li>
<li><p><strong>qual_thr</strong> (<em>int</em>) – The quality threshold
if input data come from wig this parameter is not used,</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>a function <code class="xref py py-func docutils literal notranslate"><span class="pre">get_padded_coverage()</span></code>, a function which compute the coverage for a gene on each strand
between position <em>[start, stop[</em></p>
<p>The coverage values are centered on the annot_entry.ref position, the matrix is padded by <code class="docutils literal notranslate"><span class="pre">None</span></code> value.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="o">.......</span><span class="p">[</span> <span class="n">coverage</span> <span class="n">ref</span><span class="o">.</span><span class="n">pos</span> <span class="p">]</span> <span class="o">.....</span><span class="p">]</span>
<span class="p">[</span><span class="o">....</span><span class="p">[</span><span class="n">covergae</span>     <span class="n">ref</span><span class="o">.</span><span class="n">pos</span> <span class="p">]</span> <span class="o">.....</span><span class="p">]</span>
<span class="p">[</span><span class="o">............</span><span class="p">[</span> <span class="n">cov</span> <span class="n">ref</span><span class="o">.</span><span class="n">pos</span>       <span class="p">]]</span>
</pre></div>
</div>
<p>This function take 3 parameters:</p>
<ul class="simple">
<li><p><strong>annot_entry</strong>: an entry of the annotation file.</p></li>
<li><p><strong>start</strong>: The position to start to compute the coverage(coordinates are 0-based, start position is included).</p></li>
<li><p><strong>stop</strong>: The position to stop to compute the coverage (coordinates are 0-based, stop position is excluded).</p></li>
</ul>
<p>and</p>
<ul class="simple">
<li><p><strong>return</strong>: a tuple with 2 tuple of float or int representing the coverage on strand forward and reverse.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>function</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="craw.coverage.resized_coverage_maker">
<code class="descclassname">craw.coverage.</code><code class="descname">resized_coverage_maker</code><span class="sig-paren">(</span><em>input_data</em>, <em>new_size</em>, <em>qual_thr=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/coverage.html#resized_coverage_maker"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.coverage.resized_coverage_maker" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_data</strong> (<a class="reference internal" href="wig.html#craw.wig.Genome" title="craw.wig.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">craw.wig.Genome</span></code></a> or <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.AlignmentFile</span></code> object) – the input either a samfile (see pysam library) or a genome build from a wig file (see wig module)</p></li>
<li><p><strong>new_size</strong> (<em>postive int</em>) – the number of values in the coverage vector.</p></li>
<li><p><strong>qual_thr</strong> (<em>int</em>) – The quality threshold
if input data come from wig this parameter is not used,</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>a function <code class="xref py py-func docutils literal notranslate"><span class="pre">get_resized_coverage()</span></code>, a function which compute the coverage for a gene on each strand
between position [start, stop[ This function take 3 parameters:
the coverage values are generate by linear interpolation from raw values between [start, stop[
using the scipy.
see <a class="reference external" href="https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.interpolate.interp1d.html">https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.interpolate.interp1d.html</a></p>
<ul class="simple">
<li><p><strong>annot_entry</strong>: an entry of the annotation file.</p></li>
<li><p><strong>start</strong>: The position to start to compute the coverage(coordinates are 0-based, start position is included).</p></li>
<li><p><strong>stop</strong>: The position to stop to compute the coverage (coordinates are 0-based, stop position is excluded).</p></li>
</ul>
<p>and</p>
<ul class="simple">
<li><p><strong>return</strong>: a tuple with 2 tuple of float or int representing the coverage on strand forward and reverse.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>function</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="craw.coverage.sum_coverage_maker">
<code class="descclassname">craw.coverage.</code><code class="descname">sum_coverage_maker</code><span class="sig-paren">(</span><em>input_data</em>, <em>qual_thr=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/craw/coverage.html#sum_coverage_maker"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#craw.coverage.sum_coverage_maker" title="Permalink to this definition">¶</a></dt>
<dd><p>This function return a new function <code class="xref py py-func docutils literal notranslate"><span class="pre">get_sum_coverage()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_data</strong> (<a class="reference internal" href="wig.html#craw.wig.Genome" title="craw.wig.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">craw.wig.Genome</span></code></a> or <code class="xref py py-class docutils literal notranslate"><span class="pre">pysam.AlignmentFile</span></code> object) – the input either a samfile (see pysam library) or a genome build from a wig file (see wig module)</p></li>
<li><p><strong>qual_thr</strong> (<em>int</em>) – The quality threshold
if input data come from wig this parameter is not used,</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p><code class="xref py py-func docutils literal notranslate"><span class="pre">get_sum_coverage()</span></code>, a function which compute the sum of coverage for a gene on each strand
between position [start, stop[ This function take 3 parameters:</p>
<ul class="simple">
<li><p><strong>annot_entry</strong>: an entry of the annotation file.</p></li>
<li><p><strong>start</strong>: The position to start to compute the coverage(coordinates are 0-based, start position is included).</p></li>
<li><p><strong>stop</strong>: The position to stop to compute the coverage (coordinates are 0-based, stop position is excluded).</p></li>
</ul>
<p>and</p>
<ul class="simple">
<li><p><strong>return</strong>: a tuple with 2 tuple of float or int representing the coverage on strand forward and reverse.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>function</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">coverage</a><ul>
<li><a class="reference internal" href="#functions-to-get-coverage">Functions to get coverage</a><ul>
<li><a class="reference internal" href="#get-raw-bam-coverage">get_raw_bam_coverage</a></li>
<li><a class="reference internal" href="#get-raw-wig-coverage">get_raw_wig_coverage</a></li>
<li><a class="reference internal" href="#get-raw-coverage-function">get_raw_coverage_function</a></li>
</ul>
</li>
<li><a class="reference internal" href="#functions-to-process-coverages">Functions to process coverages</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="wig.html"
                        title="previous chapter">wig</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="heatmap.html"
                        title="next chapter">heatmap</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/coverage.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="heatmap.html" title="heatmap"
             >next</a> |</li>
        <li class="right" >
          <a href="wig.html" title="wig"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017-2019, Bertrand Néron.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.0.0.
    </div>
  </body>
</html>