
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Quick start &#8212; Counter RNAseq Window 1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/style.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Inputs / Outputs" href="inputs-outputs.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="inputs-outputs.html" title="Inputs / Outputs"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="quick-start">
<span id="quickstart"></span><h1>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<div class="section" id="craw-coverage">
<h2>craw_coverage<a class="headerlink" href="#craw-coverage" title="Permalink to this headline">¶</a></h2>
<p><em>craw_coverage</em> need a file bam or wig to compute coverage and an annotation file
to specify on which regions to compute these coverages.</p>
<blockquote>
<div><ul>
<li><p>the -b or –bam allow to specify the path to the bam file.</p></li>
<li><p>or alternatively the</p>
<blockquote>
<div><ul class="simple">
<li><p>-w, –wig option to specify the path to the wig file if the both strand aer encode in same file (negative value are on reverse strand)</p></li>
<li><p>–wig-for and –wig-rev to specify the paths to the wig files for the forward and reverse strand respectively</p></li>
</ul>
</div></blockquote>
</li>
<li><p>the -a –annot allow to specify the path to the annotation file.</p></li>
</ul>
</div></blockquote>
<p>The –bam and –wig options are mutually exclusive but one of these option is required.
–wig and –wig-for or –wig-rev are also mutually exclusive.
the –annot option is required.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_coverage</span> <span class="o">--</span><span class="n">bam</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">Position</span> <span class="o">--</span><span class="n">before</span> <span class="mi">100</span> <span class="o">--</span><span class="n">after</span> <span class="mi">500</span>
<span class="n">craw_coverage</span> <span class="o">--</span><span class="n">wig</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">wig</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">Position</span> <span class="o">--</span><span class="n">before</span> <span class="mi">100</span> <span class="o">--</span><span class="n">after</span> <span class="mi">500</span>
<span class="n">craw_coverage</span> <span class="o">--</span><span class="n">wig</span><span class="o">-</span><span class="k">for</span> <span class="o">../</span><span class="n">WTE1_forward_strand</span><span class="o">.</span><span class="n">wig</span> <span class="o">--</span><span class="n">wig</span><span class="o">-</span><span class="n">rev</span> <span class="o">../</span><span class="n">WTE1_reverse_strand</span><span class="o">.</span><span class="n">wig</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> \
<span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">Position</span> <span class="o">--</span><span class="n">before</span> <span class="mi">100</span> <span class="o">--</span><span class="n">after</span> <span class="mi">500</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>At the same place of <cite>bam</cite> file, there must be the corresponding index file (the <cite>bam.bai</cite> file).
To generate the <cite>.bai</cite> file you have to use <cite>samtools</cite> program:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">samtools</span> <span class="n">index</span> <span class="n">file</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
<p>see <a class="reference external" href="http://www.htslib.org/doc/">http://www.htslib.org/doc/</a> for more explanation.</p>
</div>
<div class="section" id="with-fix-window">
<h3>with fix window<a class="headerlink" href="#with-fix-window" title="Permalink to this headline">¶</a></h3>
<p>To compute the coverage on a fix window:
we need to specify which column name in the annotation file define the reference position.
The window will computed using this reference position.</p>
<blockquote>
<div><ul class="simple">
<li><p>–ref-col</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>if –ref-col is omitted craw_coverage will use the column position. If there not “position” column
an error will occur.</p>
</div>
<p>two ways to determine the window:</p>
<p>with –window option for a window centered on the reference position.</p>
<blockquote>
<div><ul class="simple">
<li><p>–window define the number of nucleotide to take in account before and after the reference position.</p></li>
</ul>
</div></blockquote>
<a class="reference internal image-reference" href="_images/craw_window.png"><img alt="craw fixed centered windowing" src="_images/craw_window.png" style="width: 60%;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_coverage</span> <span class="o">--</span><span class="n">bam</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">Position</span> <span class="o">--</span><span class="n">window</span> <span class="mi">100</span>
</pre></div>
</div>
<p>This command will compute coverage using WTE1.bam and with annotations.txt file the column used to compute the window
is ‘Position’ and the window length will be 100 nucleotide before the reference position and 100 nucleotides after
(201 nucleotides length).</p>
<p>With an non centered window we have to specify two options <em>–before</em> and <em>–after</em></p>
<ul class="simple">
<li><p><strong>–before BEFORE</strong> define the number of nucleotide to take in account before the reference position.</p></li>
<li><p><strong>–after AFTER</strong> define the number of nucleotide to take in account after the reference position.</p></li>
</ul>
<a class="reference internal image-reference" href="_images/craw_window_before_after.png"><img alt="craw fixed non centered windowing" src="_images/craw_window_before_after.png" style="width: 60%;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_coverage</span> <span class="o">--</span><span class="n">bam</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">Position</span> <span class="o">--</span><span class="n">before</span> <span class="mi">100</span> <span class="o">--</span><span class="n">after</span> <span class="mi">500</span>
</pre></div>
</div>
<p>This command will compute coverage using WTE1.bam and with annotations.txt file the column used to compute the window
is ‘Position’ and the window length will be 100 nucleotide before the reference position and 500 nucleotides after
(201 nucleotides length).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>–after and –before options must be set together and are incompatible with –window option.</p>
</div>
</div>
<div class="section" id="with-variable-window">
<h3>with variable window<a class="headerlink" href="#with-variable-window" title="Permalink to this headline">¶</a></h3>
<p>The regions must be specified in the annotation file.</p>
<ul class="simple">
<li><p><strong>–start-col COL</strong> define the name of the column in annotation file which define the start position of the region to compute.</p></li>
<li><p><strong>–stop-col COL</strong> define the name of the column in annotation file which define the stop position of the region to compute.</p></li>
</ul>
<a class="reference internal image-reference" href="_images/craw_window_variable.png"><img alt="craw variable windowing" src="_images/craw_window_variable.png" style="width: 60%;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_coverage</span> <span class="o">--</span><span class="n">bam</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_start</span> <span class="o">--</span><span class="n">start</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_start</span>  <span class="o">--</span><span class="n">stop</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_end</span>
</pre></div>
</div>
<p>This command will compute coverage using WTE1.bam and with annotations.txt file.</p>
<ul class="simple">
<li><p>The reference position will define by the <em>annotation_start</em> column</p></li>
<li><p>The first nucleotide of the window will be define by <em>annotation_start</em> column.</p></li>
<li><p>The last nucleotide of the window will be define by <em>annotation_end</em> column.</p></li>
</ul>
<p>By default the coverages are aligned using the column col-ref as reference, if each window have not the same size
(see above) the <code class="docutils literal notranslate"><span class="pre">None</span></code> value is used to padded the coverage matrix.
There is 2 other options which alter this behavior</p>
</div>
<div class="section" id="justify-option">
<h3>justify option<a class="headerlink" href="#justify-option" title="Permalink to this headline">¶</a></h3>
<p>When this option is set. <code class="docutils literal notranslate"><span class="pre">craw_coverage</span></code> get the coverages between the position start and stop and then generate new values
by linear interpolation to have the same number of coverage measure for all genes.
If this option is used with fix window and the window is out the gene (before the first position) only the position from
0 will take in account to generates new values.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_coverage</span> <span class="o">--</span><span class="n">bam</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_start</span> <span class="o">--</span><span class="n">start</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_start</span>  <span class="o">--</span><span class="n">stop</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_end</span> <span class="o">--</span><span class="n">justify</span> <span class="mi">1000</span>
</pre></div>
</div>
<p>This command will compute coverage using WTE1.bam and with annotations.txt file.
For each gene, 1000 values will be estimated by linear interpolation from the raw coverage values.</p>
</div>
<div class="section" id="sum-option">
<h3>sum option<a class="headerlink" href="#sum-option" title="Permalink to this headline">¶</a></h3>
<p>When this option is set, instead of recording the coverage for each nucleotides, <code class="docutils literal notranslate"><span class="pre">craw_coverage</span></code> compute
the sum of the coverages of the window and report it on the coverage file.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_coverage</span> <span class="o">--</span><span class="n">bam</span> <span class="o">../</span><span class="n">WTE1</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">annot</span> <span class="o">../</span><span class="n">annotations</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_start</span> <span class="o">--</span><span class="n">start</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_start</span>  <span class="o">--</span><span class="n">stop</span><span class="o">-</span><span class="n">col</span> <span class="n">annotation_end</span> <span class="o">--</span><span class="nb">sum</span>
</pre></div>
</div>
<p>This command will compute coverage using WTE1.bam and with annotations.txt file.
For each gene, the sum of all coverages is computed and reported on the coverage file.</p>
</div>
<div class="section" id="other-options">
<h3>Other options<a class="headerlink" href="#other-options" title="Permalink to this headline">¶</a></h3>
<p>The following option are not mandatory:</p>
<ul class="simple">
<li><p><strong>-q QUAL_THR, –qual-thr QUAL_THR</strong> The minimal quality of read mapping to take it in account. (default=15)</p></li>
<li><p><strong>-s SUFFIX, –suffix SUFFIX</strong> The name of the suffix to use for the output file. (default= <cite>.cov</cite>)</p></li>
<li><p><strong>-o OUTPUT, –output OUTPUT</strong> The path of the output (default= base name of annotation file with –suffix)</p></li>
<li><p><strong>–version</strong> display version information and quit.</p></li>
<li><p><strong>–verbose, -v</strong> increase the verbosity of the output (this option can be repeat several times as -vv).</p></li>
<li><p><strong>–quiet</strong> decrease verbosity of the output. By default craw_coverage is slightly verbose and display a progress
bar. This option can be useful to disable any progression information on batch run.</p></li>
<li><p><strong>-h –help</strong> display the inline help and exit.</p></li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>by default craw_coverage use a quality threshold of 15 (like pysam)</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>strand column mut named <em>strand</em> and can take <cite>1/-1</cite> or <cite>+/-</cite> <cite>for/rev</cite> as value for forward/reverse strands.</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>the coverage file can be huge depending on the number of gene to compute the coverage and the size of the window
for instance for 6000 genes with a window of 15000 nt the cov file will weight almost 900Mb.</p>
</div>
</div>
</div>
<div class="section" id="craw-htmp">
<h2>craw_htmp<a class="headerlink" href="#craw-htmp" title="Permalink to this headline">¶</a></h2>
<p>Compute a figure from a file of coverage generated by <cite>craw_coverage</cite>.
By default, display a figure with two heatmap one for the sense the other for the antisense.
But it work also if the coverage file contains <em>sense</em> or <em>anti sense</em> data only.</p>
<div class="section" id="mandatory-arguments">
<h3>Mandatory arguments<a class="headerlink" href="#mandatory-arguments" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>cov_file</strong> The path to the coverage file ( the output of ).</p></li>
</ul>
</div>
<div class="section" id="data-options">
<h3>Data options<a class="headerlink" href="#data-options" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p><strong>–crop CROP CROP</strong>: Crop the matrix. This option need two values the name of the first and last column to keep
[start col, stop col] eg –crop -10 1000</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_htmp</span> <span class="o">--</span><span class="n">crop</span> <span class="mi">0</span> <span class="mi">2000</span> <span class="n">WTE1_var_window</span><span class="o">.</span><span class="n">cov</span>
</pre></div>
</div>
<p>This command will display only column ‘0’ to ‘2000’, included, of the matrix generated by craw_coverage.</p>
</li>
<li><p><strong>–sort-using-col COL</strong> sort the data using the column name ‘COL’ (descending).</p></li>
<li><p><strong>–sort-using-file SORT_USING_FILE</strong> sort the rows using a file.
The file must have on the first line the name of the column to use for sorting
and each line must match to a value contained in the matrix.</p></li>
<li><p><strong>–sort-by-gene-size [start_col,stop_col [start_col,stop_col …]]</strong>
The rows will be sorted by gene size using <em>start_col</em> and <em>stop_col</em> to compute length.
<em>start_col</em> and <em>stop_col</em> must be a string separated by comma.
If start_col and stop_col are not specify <em>annotation_start,annotation_end</em> will be used.</p></li>
<li><p><strong>–sense-only</strong>          Display only sense matrix (default is display both).</p></li>
<li><p><strong>–antisense-only</strong>      Display only anti sense matrix (default is display both).</p></li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Don’t put the  <strong>–sort-by-gene-size</strong> option without value as last option just before the coverage file.
In this case the craw_htmp will don’t work. If you want to use only this option, use the <strong>-v</strong> option after
<strong>–sort-by-gene-size</strong></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_htmp</span> <span class="o">--</span><span class="n">sort</span><span class="o">-</span><span class="n">by</span><span class="o">-</span><span class="n">gene</span><span class="o">-</span><span class="n">size</span> <span class="o">-</span><span class="n">v</span> <span class="n">WTE1_0_2000</span><span class="o">.</span><span class="n">cov</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="figure-options">
<h3>Figure options<a class="headerlink" href="#figure-options" title="Permalink to this headline">¶</a></h3>
<div class="section" id="normalization-options">
<h4>Normalization options<a class="headerlink" href="#normalization-options" title="Permalink to this headline">¶</a></h4>
<p><strong>craw_htmp</strong> provide several methods to normalize data before to display them:
below the different figures illustrate the result of each normalization methods on
the following matrix <span class="math notranslate nohighlight">\(x\)</span>.</p>
<table class="docutils align-center">
<colgroup>
<col style="width: 12%" />
<col style="width: 17%" />
<col style="width: 15%" />
<col style="width: 12%" />
<col style="width: 12%" />
<col style="width: 15%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Pos</p></th>
<th class="head"><p>0</p></th>
<th class="head"><p>1</p></th>
<th class="head"><p>2</p></th>
<th class="head"><p>3</p></th>
<th class="head"><p>4</p></th>
<th class="head"><p>5</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>S</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
<td><p>10</p></td>
<td><p>100</p></td>
<td><p>1000</p></td>
<td><p>10000</p></td>
</tr>
<tr class="row-odd"><td><p>AS</p></td>
<td><p>10000</p></td>
<td><p>1000</p></td>
<td><p>100</p></td>
<td><p>10</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>S</p></td>
<td><p>1</p></td>
<td><p>10</p></td>
<td><p>20</p></td>
<td><p>30</p></td>
<td><p>40</p></td>
<td><p>50</p></td>
</tr>
<tr class="row-odd"><td><p>AS</p></td>
<td><p>50</p></td>
<td><p>40</p></td>
<td><p>30</p></td>
<td><p>20</p></td>
<td><p>10</p></td>
<td><p>1</p></td>
</tr>
</tbody>
</table>
<div class="section" id="linear-normalisation">
<h5>linear normalisation<a class="headerlink" href="#linear-normalisation" title="Permalink to this headline">¶</a></h5>
<div class="docutils container">
<a class="reference internal image-reference" href="_images/norm_demo_lin.png"><img alt="linear normalization" class="align-right" src="_images/norm_demo_lin.png" style="width: 25%;" /></a>
<p><strong>–norm lin</strong>: a linear normalization is applied on the whole matrix.</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(x\)</span> the original matrix</p></li>
<li><p><span class="math notranslate nohighlight">\(x_i\)</span> the value of a cell in the original matrix</p></li>
<li><p><span class="math notranslate nohighlight">\(z_i\)</span> the value of a cell in the normalized matrix</p></li>
<li><p><span class="math notranslate nohighlight">\(z_i = \frac{x_i - min(x)}{max(x) - min(x)}\)</span></p></li>
</ul>
</div>
<div class="clearer docutils container">
<img alt="_images/spacer.png" src="_images/spacer.png" />
</div>
</div>
<div class="section" id="logaritmic-normalisation">
<h5>logaritmic normalisation<a class="headerlink" href="#logaritmic-normalisation" title="Permalink to this headline">¶</a></h5>
<div class="docutils container">
<a class="reference internal image-reference" href="_images/norm_demo_log.png"><img alt="logaritmic normalization" class="align-right" src="_images/norm_demo_log.png" style="width: 25%;" /></a>
<p><strong>–norm log</strong>: a 10 base logarithm will be applied on the data before matrix normalization.</p>
<ol class="arabic simple">
<li><p>replace all 0 values by 1</p></li>
<li><p><span class="math notranslate nohighlight">\(z_i = log_{10}(x_i)\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(w_i = \frac{z_i - min(z)}{max(z) - min(z)}\)</span></p></li>
</ol>
</div>
<div class="clearer docutils container">
<img alt="_images/spacer.png" src="_images/spacer.png" />
</div>
</div>
<div class="section" id="linear-normalisation-row-by-row">
<h5>linear normalisation row by row<a class="headerlink" href="#linear-normalisation-row-by-row" title="Permalink to this headline">¶</a></h5>
<div class="docutils container">
<a class="reference internal image-reference" href="_images/norm_demo_row.png"><img alt="linear normalization row by row" class="align-right" src="_images/norm_demo_row.png" style="width: 25%;" /></a>
<p><strong>–norm row</strong>: mean that a linear normalisation is compute row by row.</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(x\)</span> the original matrix</p></li>
<li><p><span class="math notranslate nohighlight">\(x_{ij}\)</span> the value of a cell in the original matrix with <span class="math notranslate nohighlight">\(I\)</span> rows and <span class="math notranslate nohighlight">\(J\)</span> columns</p></li>
<li><p><span class="math notranslate nohighlight">\(x_i\)</span> the values of the <span class="math notranslate nohighlight">\(i\)</span> row</p></li>
<li><p><span class="math notranslate nohighlight">\(z_{ij} = \frac{x_{ij} - min(x_i)}{max(x_i) - min(x_i)}\)</span></p></li>
</ul>
</div>
<div class="clearer docutils container">
<img alt="_images/spacer.png" src="_images/spacer.png" />
</div>
</div>
<div class="section" id="logaritmic-normalisation-row-by-row">
<h5>logaritmic normalisation row by row<a class="headerlink" href="#logaritmic-normalisation-row-by-row" title="Permalink to this headline">¶</a></h5>
<div class="docutils container">
<a class="reference internal image-reference" href="_images/norm_demo_row+log.png"><img alt="logaritmic normalization row by row" class="align-right" src="_images/norm_demo_row+log.png" style="width: 25%;" /></a>
<p><strong>–norm log+row</strong> mean a 10 base logarithm will be applied before a normalisation row by row.</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(x\)</span> the original matrix</p></li>
<li><p><span class="math notranslate nohighlight">\(x_{ij}\)</span> the value of a cell in the original matrix with <span class="math notranslate nohighlight">\(I\)</span> rows and <span class="math notranslate nohighlight">\(J\)</span> columns</p></li>
<li><p><span class="math notranslate nohighlight">\(x_i\)</span> the values of the <span class="math notranslate nohighlight">\(i\)</span> row</p></li>
</ul>
<ol class="arabic simple">
<li><p>replace all 0 values by 1</p></li>
<li><p><span class="math notranslate nohighlight">\(z_{ij} = log_{10}(x_ij)\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(w_{ij} = \frac{z_{ij} - min(z_i)}{max(z_i) - min(z_i)}\)</span></p></li>
</ol>
</div>
<div class="clearer docutils container">
<img alt="_images/spacer.png" src="_images/spacer.png" />
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><strong>‘row+log’</strong> is an alias for <strong>‘log+row’</strong></p></li>
<li><p>The default normalisation is <strong>lin</strong></p></li>
</ul>
</div>
</div>
</div>
<div class="section" id="other-figure-options">
<h4>Other figure options<a class="headerlink" href="#other-figure-options" title="Permalink to this headline">¶</a></h4>
<ul>
<li><p><strong>–cmap CMAP</strong> The color map used to display data. The allowed values are defined in
<a class="reference external" href="http:matplotlib.org/examples/color/colormaps_reference.html">http:matplotlib.org/examples/color/colormaps_reference.html</a>
eg: Blues, BuGn, Greens, GnBu, … (default: Blues).</p></li>
<li><p><strong>–title TITLE</strong> The figure title. It will display on the top of the figure.
(default: the name of the coverage file without extension).</p></li>
<li><p><strong>–dpi DPI</strong> The resolution of the output (default=100).</p>
<p>This option work only if <strong>–out</strong> option is specified. <br />
To set the right dpi for screen displaying use the <a class="reference internal" href="installation.html#matplotlibrc"><span class="std std-ref">matplotlib configuration</span></a> file.</p>
</li>
<li><p><strong>–size SIZE</strong> Specify the figure size</p>
<p>The value must be widexheight[unit] or ‘raw’.
If value is ‘raw’ it will be produce two image files (for sense and antisense)
with one pixel correspond to one coverage value.
Otherwise, ‘wide’ and ‘height’ must be positive integers
By default <em>unit</em> is in inches. eg:</p>
<ul class="simple">
<li><p>7x10 or 7x10in for 7 inches wide by 10 inches height.</p></li>
<li><p>70x100mm for 70 mm by 100 mm.</p></li>
</ul>
<p>default=7x10 or 10x7 depending of the figure orientation (see layout).</p>
</li>
<li><p><strong>–mark</strong> POS &lt;COLOR&gt; will draw a vertical line at the position POS with the color &lt;COLOR&gt;</p>
<div class="docutils container">
<a class="reference internal image-reference" href="_images/wte1_0+2000_mark.png"><img alt="coverage matrix with marks" class="align-right" src="_images/wte1_0+2000_mark.png" style="width: 25%;" /></a>
<p>COLOR can be the name of the most common html color red, yellow, … or a value of a RGB in hexadecimal format
like #rgb or #rrggbb for instance #ff0000 represent pure red.
(don’t forget to surround the hexadecimal color with  quotes on commandline)</p>
<p>If COLOR is omitted the color of the highest value of the color map used for the drawing will be used
(The default color map is Blues).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_htmp</span> <span class="o">--</span><span class="n">norm</span> <span class="n">log</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">1000</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">500</span> <span class="n">MediumAquamarine</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">100</span> <span class="s2">&quot;#ff0000&quot;</span> <span class="o">--</span><span class="n">sense</span><span class="o">-</span><span class="n">only</span> <span class="n">WTE1_0</span><span class="o">+</span><span class="mf">2000.</span><span class="n">cov</span>
</pre></div>
</div>
<p>for list of HTML colors:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Web_colors">https://en.wikipedia.org/wiki/Web_colors</a></p></li>
<li><p><a class="reference external" href="https://www.w3schools.com/colors/colors_names.asp">https://www.w3schools.com/colors/colors_names.asp</a></p></li>
</ul>
</div>
<div class="clearer docutils container">
<img alt="_images/spacer.png" src="_images/spacer.png" />
</div>
</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The –mark option must not be the last option on the command line (just before the coverage file),
otherwise an error will occurred.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">craw_htmp</span> <span class="o">--</span><span class="n">out</span> <span class="n">my_fig</span><span class="o">.</span><span class="n">png</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">10</span> <span class="n">red</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">0</span>  <span class="n">WTE1_0_2000</span><span class="o">.</span><span class="n">cov</span> <span class="o">=&gt;</span> <span class="k">raise</span> <span class="n">an</span> <span class="n">error</span>
<span class="n">craw_htmp</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">10</span> <span class="n">red</span> <span class="o">--</span><span class="n">mark</span> <span class="mi">0</span> <span class="o">--</span><span class="n">out</span> <span class="n">my_fig</span><span class="o">.</span><span class="n">png</span> <span class="n">WTE1_0_2000</span><span class="o">.</span><span class="n">cov</span> <span class="o">=&gt;</span> <span class="n">work</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="layout-options">
<h4>Layout options<a class="headerlink" href="#layout-options" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><strong>–sense-on-left</strong>       Where to display the sense matrix relative to antisense matrix.</p></li>
<li><p><strong>–sense-on-right</strong>      Where to display the sense matrix relative to antisense matrix.</p></li>
<li><p><strong>–sense-on-top</strong>        Where to display the sense matrix relative to antisense matrix.</p></li>
<li><p><strong>–sense-on-bottom</strong>     Where to display the sense matrix relative to antisense matrix.</p></li>
</ul>
<div class="docutils container">
<a class="reference internal image-reference" href="_images/craw_htmp_sense_on_top.png"><img alt="--sense-on-top" src="_images/craw_htmp_sense_on_top.png" style="width: 20%;" /></a>
<a class="reference internal image-reference" href="_images/craw_htmp_sense_on_left.png"><img alt="--sense-on-left" src="_images/craw_htmp_sense_on_left.png" style="width: 30%;" /></a>
</div>
<p>The fisrt screen capture use <em>–sense-on-top</em> whereas the second capture used <em>–sense-on-left option</em>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>default is top.</p>
</div>
</div>
</div>
<div class="section" id="id1">
<h3>Other options<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><strong>-h, –help</strong>  Display the help message and exit</p></li>
<li><p><strong>–out OUT</strong>  The name of the file (the format will based on the extension) to save the figure.
Instead of displaying the figure on the screen, save it directly in this file.</p></li>
<li><p><strong>-v, –verbose</strong> Increase output verbosity. By default craw_htmp is relatively quiet (display only warning and error),
if you want to display also the processing step just add -v on the commandline (or -vv to display also the debugging message).</p></li>
<li><p><strong>–version</strong> Display version information and quit.</p></li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Quick start</a><ul>
<li><a class="reference internal" href="#craw-coverage">craw_coverage</a><ul>
<li><a class="reference internal" href="#with-fix-window">with fix window</a></li>
<li><a class="reference internal" href="#with-variable-window">with variable window</a></li>
<li><a class="reference internal" href="#justify-option">justify option</a></li>
<li><a class="reference internal" href="#sum-option">sum option</a></li>
<li><a class="reference internal" href="#other-options">Other options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#craw-htmp">craw_htmp</a><ul>
<li><a class="reference internal" href="#mandatory-arguments">Mandatory arguments</a></li>
<li><a class="reference internal" href="#data-options">Data options</a></li>
<li><a class="reference internal" href="#figure-options">Figure options</a><ul>
<li><a class="reference internal" href="#normalization-options">Normalization options</a><ul>
<li><a class="reference internal" href="#linear-normalisation">linear normalisation</a></li>
<li><a class="reference internal" href="#logaritmic-normalisation">logaritmic normalisation</a></li>
<li><a class="reference internal" href="#linear-normalisation-row-by-row">linear normalisation row by row</a></li>
<li><a class="reference internal" href="#logaritmic-normalisation-row-by-row">logaritmic normalisation row by row</a></li>
</ul>
</li>
<li><a class="reference internal" href="#other-figure-options">Other figure options</a></li>
<li><a class="reference internal" href="#layout-options">Layout options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id1">Other options</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="inputs-outputs.html"
                        title="next chapter">Inputs / Outputs</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/quickstart.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="inputs-outputs.html" title="Inputs / Outputs"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Counter RNAseq Window 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017-2019, Bertrand Néron.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.0.0.
    </div>
  </body>
</html>